{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2a552949",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "00f1732b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>info</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14412</td>\n",
       "      <td>+</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>DDX11L1-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>DDX11L1-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>+</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>DDX11L1-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>DDX11L1-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2619439</th>\n",
       "      <td>chrM</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>+</td>\n",
       "      <td>gene_id \"ENSG00000210195.2\"; transcript_id \"EN...</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TT-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2619440</th>\n",
       "      <td>chrM</td>\n",
       "      <td>exon</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>+</td>\n",
       "      <td>gene_id \"ENSG00000210195.2\"; transcript_id \"EN...</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TT-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2619441</th>\n",
       "      <td>chrM</td>\n",
       "      <td>gene</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000210196.2\"; transcript_id \"EN...</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2619442</th>\n",
       "      <td>chrM</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000210196.2\"; transcript_id \"EN...</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TP-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2619443</th>\n",
       "      <td>chrM</td>\n",
       "      <td>exon</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000210196.2\"; transcript_id \"EN...</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TP-201</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2619444 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom        type  start    end strand  \\\n",
       "0        chr1        gene  11869  14412      +   \n",
       "1        chr1  transcript  11869  14409      +   \n",
       "2        chr1        exon  11869  12227      +   \n",
       "3        chr1        exon  12613  12721      +   \n",
       "4        chr1        exon  13221  14409      +   \n",
       "...       ...         ...    ...    ...    ...   \n",
       "2619439  chrM  transcript  15888  15953      +   \n",
       "2619440  chrM        exon  15888  15953      +   \n",
       "2619441  chrM        gene  15956  16023      -   \n",
       "2619442  chrM  transcript  15956  16023      -   \n",
       "2619443  chrM        exon  15956  16023      -   \n",
       "\n",
       "                                                      info gene_name  \\\n",
       "0        gene_id \"ENSG00000223972.4\"; transcript_id \"EN...   DDX11L1   \n",
       "1        gene_id \"ENSG00000223972.4\"; transcript_id \"EN...   DDX11L1   \n",
       "2        gene_id \"ENSG00000223972.4\"; transcript_id \"EN...   DDX11L1   \n",
       "3        gene_id \"ENSG00000223972.4\"; transcript_id \"EN...   DDX11L1   \n",
       "4        gene_id \"ENSG00000223972.4\"; transcript_id \"EN...   DDX11L1   \n",
       "...                                                    ...       ...   \n",
       "2619439  gene_id \"ENSG00000210195.2\"; transcript_id \"EN...        MT   \n",
       "2619440  gene_id \"ENSG00000210195.2\"; transcript_id \"EN...        MT   \n",
       "2619441  gene_id \"ENSG00000210196.2\"; transcript_id \"EN...        MT   \n",
       "2619442  gene_id \"ENSG00000210196.2\"; transcript_id \"EN...        MT   \n",
       "2619443  gene_id \"ENSG00000210196.2\"; transcript_id \"EN...        MT   \n",
       "\n",
       "        transcript_name  \n",
       "0               DDX11L1  \n",
       "1           DDX11L1-002  \n",
       "2           DDX11L1-002  \n",
       "3           DDX11L1-002  \n",
       "4           DDX11L1-002  \n",
       "...                 ...  \n",
       "2619439       MT-TT-201  \n",
       "2619440       MT-TT-201  \n",
       "2619441           MT-TP  \n",
       "2619442       MT-TP-201  \n",
       "2619443       MT-TP-201  \n",
       "\n",
       "[2619444 rows x 8 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v19_anno = pd.read_csv('/lmh_data/data/sclab/public/gencode.v19.annotation.gtf',\n",
    "                       header=None, sep='\\t', skiprows=[i for i in range(5)], usecols=[0, 2, 3, 4, 6, 8], names=[\n",
    "                           'chrom', 'type', 'start', 'end', 'strand', 'info'\n",
    "                       ])\n",
    "v19_anno['gene_name'] = v19_anno['info'].str.extract(r'(gene_name \")(\\w*)')[1]\n",
    "v19_anno['transcript_name'] = v19_anno['info'].str.extract(r'(transcript_name \")(\\S*)(\")')[1]\n",
    "v19_anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "50c254bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>info</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14412</td>\n",
       "      <td>+</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>DDX11L1-002</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>DDX11L1-002</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>+</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>DDX11L1-002</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>DDX11L1-002</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2619439</th>\n",
       "      <td>chrM</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>+</td>\n",
       "      <td>gene_id \"ENSG00000210195.2\"; transcript_id \"EN...</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TT-201</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2619440</th>\n",
       "      <td>chrM</td>\n",
       "      <td>exon</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>+</td>\n",
       "      <td>gene_id \"ENSG00000210195.2\"; transcript_id \"EN...</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TT-201</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2619441</th>\n",
       "      <td>chrM</td>\n",
       "      <td>gene</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000210196.2\"; transcript_id \"EN...</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TP</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2619442</th>\n",
       "      <td>chrM</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000210196.2\"; transcript_id \"EN...</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TP-201</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2619443</th>\n",
       "      <td>chrM</td>\n",
       "      <td>exon</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000210196.2\"; transcript_id \"EN...</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TP-201</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2619444 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom        type  start    end strand  \\\n",
       "0        chr1        gene  11869  14412      +   \n",
       "1        chr1  transcript  11869  14409      +   \n",
       "2        chr1        exon  11869  12227      +   \n",
       "3        chr1        exon  12613  12721      +   \n",
       "4        chr1        exon  13221  14409      +   \n",
       "...       ...         ...    ...    ...    ...   \n",
       "2619439  chrM  transcript  15888  15953      +   \n",
       "2619440  chrM        exon  15888  15953      +   \n",
       "2619441  chrM        gene  15956  16023      -   \n",
       "2619442  chrM  transcript  15956  16023      -   \n",
       "2619443  chrM        exon  15956  16023      -   \n",
       "\n",
       "                                                      info gene_name  \\\n",
       "0        gene_id \"ENSG00000223972.4\"; transcript_id \"EN...   DDX11L1   \n",
       "1        gene_id \"ENSG00000223972.4\"; transcript_id \"EN...   DDX11L1   \n",
       "2        gene_id \"ENSG00000223972.4\"; transcript_id \"EN...   DDX11L1   \n",
       "3        gene_id \"ENSG00000223972.4\"; transcript_id \"EN...   DDX11L1   \n",
       "4        gene_id \"ENSG00000223972.4\"; transcript_id \"EN...   DDX11L1   \n",
       "...                                                    ...       ...   \n",
       "2619439  gene_id \"ENSG00000210195.2\"; transcript_id \"EN...        MT   \n",
       "2619440  gene_id \"ENSG00000210195.2\"; transcript_id \"EN...        MT   \n",
       "2619441  gene_id \"ENSG00000210196.2\"; transcript_id \"EN...        MT   \n",
       "2619442  gene_id \"ENSG00000210196.2\"; transcript_id \"EN...        MT   \n",
       "2619443  gene_id \"ENSG00000210196.2\"; transcript_id \"EN...        MT   \n",
       "\n",
       "        transcript_name       transcript_type  \n",
       "0               DDX11L1            pseudogene  \n",
       "1           DDX11L1-002  processed_transcript  \n",
       "2           DDX11L1-002  processed_transcript  \n",
       "3           DDX11L1-002  processed_transcript  \n",
       "4           DDX11L1-002  processed_transcript  \n",
       "...                 ...                   ...  \n",
       "2619439       MT-TT-201               Mt_tRNA  \n",
       "2619440       MT-TT-201               Mt_tRNA  \n",
       "2619441           MT-TP               Mt_tRNA  \n",
       "2619442       MT-TP-201               Mt_tRNA  \n",
       "2619443       MT-TP-201               Mt_tRNA  \n",
       "\n",
       "[2619444 rows x 9 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v19_anno['transcript_type'] = v19_anno['info'].str.extract(r'(transcript_type \")(\\S*)(\")')[1]\n",
    "v19_anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1e323bc7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['pseudogene', 'processed_transcript',\n",
       "       'transcribed_unprocessed_pseudogene', 'unprocessed_pseudogene',\n",
       "       'lincRNA', 'miRNA', 'protein_coding', 'processed_pseudogene',\n",
       "       'antisense', 'snRNA', 'retained_intron', 'nonsense_mediated_decay',\n",
       "       'sense_intronic', 'misc_RNA', 'transcribed_processed_pseudogene',\n",
       "       'snoRNA', 'non_stop_decay', 'rRNA', 'unitary_pseudogene',\n",
       "       '3prime_overlapping_ncrna', 'polymorphic_pseudogene',\n",
       "       'sense_overlapping', 'IG_V_gene', 'IG_C_gene', 'IG_J_gene',\n",
       "       'IG_V_pseudogene', 'TR_C_gene', 'TR_J_gene', 'TR_V_gene',\n",
       "       'TR_V_pseudogene', 'IG_C_pseudogene', 'TR_D_gene',\n",
       "       'TR_J_pseudogene', 'translated_processed_pseudogene',\n",
       "       'IG_J_pseudogene', 'IG_D_gene', 'Mt_tRNA', 'Mt_rRNA'], dtype=object)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v19_anno['transcript_type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b157a4f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def test(genes):\n",
    "    filter_genes_info = v19_anno[\n",
    "        (v19_anno['transcript_type']=='protein_coding')&\n",
    "        (v19_anno['gene_name'].isin(genes))&\n",
    "        (v19_anno['type']=='gene')\n",
    "    ]\n",
    "    print(filter_genes_info.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f113db47",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(97, 9)\n"
     ]
    }
   ],
   "source": [
    "# PDGFRA_body\n",
    "test(['KCNIP4', 'MALAT1', 'LHFPL3', 'LRRC4C', 'CSMD1', 'PCDH9', 'ERBB4', 'NRXN1', 'DMD', 'IL1RAPL1', 'XIST', 'LSAMP', 'SOX2-OT', 'GRIA2', 'GRM5', 'TNR', 'PTPRG', 'PDZD2', 'GRIA4', 'NTM', 'GALNT13', 'NPAS3', 'SHISA9', 'MARCH1', 'NTRK2', 'PHF14', 'ZSWIM6', 'PTPRZ1', 'GPM6A', 'NFIB', 'PCDH15', 'ANKS1B', 'SOX5', 'NEGR1', 'TMEM108', 'ANK2', 'SLC8A1', 'SPOCK3', 'PDE4B', 'CALM2', 'FGF14', 'LPPR1', 'MAPK10', 'NAV1', 'IL1RAP', 'FHIT', 'NLGN4X', 'PDE4D', 'KAZN', 'FAM155A', 'GPC6', 'DOCK4', 'COL9A1', 'PTPRD', 'ZNF521', 'CPE', 'OLIG1', 'SEZ6L', 'ANK3', 'MAGI2', 'LRFN5', 'DNER', 'GLCCI1', 'PSD3', 'ASTN2', 'DCAF6', 'SORCS1', 'FGF12', 'XYLT1', 'GRID2', 'RGS7', 'TJP1', 'ALCAM', 'NCAM2', 'CTTNBP2', 'GRM7', 'FTH1', 'GUCY1A2', 'FAF1', 'PLXDC2', 'SEMA6A', 'FBXL17', 'OPCML', 'ZNF638', 'MEIS2', 'NOL4', 'TCF12', 'ZFPM2', 'DNM3', 'PPFIA2', 'AGAP1', 'NLGN1', 'LPP', 'ADARB2', 'ZHX3', 'PHACTR3', 'SIK3', 'CCSER1', 'GSK3B', 'PRTFDC1'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3fe3071f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(97, 9)\n"
     ]
    }
   ],
   "source": [
    "# PDGFRA_promoter\n",
    "test(['KCNIP4', 'LHFPL3', 'MALAT1', 'NRXN1', 'CSMD1', 'PCDH9', 'LRRC4C', 'ERBB4', 'PCDH15', 'LSAMP', 'XIST', 'TNR', 'DMD', 'IL1RAPL1', 'GRIA2', 'NPAS3', 'NTM', 'MARCH1', 'GRM5', 'SOX2-OT', 'GALNT13', 'GRIA4', 'NTRK2', 'NFIB', 'GPM6A', 'ANK2', 'PTPRZ1', 'ANKS1B', 'ASTN2', 'FGF14', 'NEGR1', 'SHISA9', 'KAZN', 'DOCK4', 'MAGI2', 'ZSWIM6', 'SLC8A1', 'LPPR1', 'SPOCK3', 'SOX5', 'PDZD2', 'IL1RAP', 'PTPRD', 'PDE4B', 'GPC6', 'NLGN1', 'ZNF521', 'FAM155A', 'PDE4D', 'OPCML', 'NLGN4X', 'CPE', 'CLASP2', 'FHIT', 'SCD5', 'KCNMB2', 'PHF14', 'GRM7', 'RGS7', 'DNER', 'SEZ6L', 'HIP1', 'PPP1R12A', 'SLC44A5', 'TMEM132C', 'MAPK10', 'GRID2', 'TMEM108', 'TJP1', 'SORCS1', 'RFTN2', 'GSK3B', 'CTTNBP2', 'VCAN', 'NCAM2', 'N4BP2L2', 'PSD3', 'ANKRD17', 'PREX2', 'COL9A1', 'SIK3', 'SORBS2', 'JMJD1C', 'PLXDC2', 'CCSER1', 'PPFIA2', 'AGAP1', 'DPP10', 'PTPRG', 'RIT2', 'ZNF638', 'NFIA', 'XYLT1', 'SNTG1', 'CPEB4', 'LRFN5', 'KLHL13', 'ZHX3', 'CST3', 'RIN2'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "af6a0cd4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(22, 9)\n"
     ]
    }
   ],
   "source": [
    "# PDGFRA_different\n",
    "test(['RIN2', 'TMEM132C', 'KCNMB2', 'HIP1', 'CLASP2', 'JMJD1C', 'RFTN2', 'ANKRD17', 'SLC44A5', 'PPP1R12A', 'SCD5', 'SNTG1', 'VCAN', 'PREX2', 'N4BP2L2', 'RIT2', 'NFIA', 'SORBS2', 'DPP10', 'CST3', 'CPEB4', 'KLHL13'])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sclab",
   "language": "python",
   "name": "sclab"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
